common_crawler/crawler.py

```diff
@@ -79,21 +79,35 @@ def search_common_crawl_index(self, url: str, page: int = 0) -> list[dict]:
         search_url.add_parameter('output', 'json')
         search_url.add_parameter('page', page)
 
-        # Perform an HTTP GET request to retrieve records for the encoded URL.
-        response = requests.get(str(search_url))
-
-        # If the request is successful, parse each record from the response and return them.
-        if response.status_code == 200:
-            records = response.text.strip().split('\n')
-            print(f"Found {len(records)} records for {url} on page {page}")
-            return [json.loads(record) for record in records]
-        elif 'First Page is 0, Last Page is 0' in response.text:
-            print("No records exist in index matching the url search term")
-            return None
-        else:
-            print(f"Failed to get records for {url} on page {page}: {response.text}")
-            # Return None to indicate that no records were found or an error occurred.
-            return None
+        retries = 0
+        delay = 1
+
+        # put HTTP GET request in re-try loop in case of rate limiting. Once per second is nice enough per common crawl doc.
+        while retries < max_retries:
+            try:
+                # Perform an HTTP GET request to retrieve records for the encoded URL.
+                response = requests.get(str(search_url))
+                response.raise_for_status()
+
+                # If the request is successful, parse each record from the response and return them.
+                if response.status_code == 200:
+                    records = response.text.strip().split('\n')
+                    print(f"Found {len(records)} records for {url} on page {page}")
+                    return [json.loads(record) for record in records]
+                elif 'First Page is 0, Last Page is 0' in response.text:
+                    print("No records exist in index matching the url search term")
+                    return None
+            except requests.exceptions.RequestException as e:
+                if response.status_code == 500 and 'SlowDown' in response.text:
+                    retries += 1
+                    print(f"Rate limit exceeded. Rerun in {delay} second(s)... (Attempt {retries}/{max_retries})")
+                    time.sleep(delay)
+                else:
+                    print(f"Failed to get records for {url} on page {page}: {response.text}")
+                    return None
```

Comment:
This section of the code is three indentations at its deepest: an if-else within a try-except within a while. Once we get to this part of the code, it's best to extract at least part of the function so as to avoid confusion. Otherwise, we fall into the dreaded [arrowhead anti-pattern](https://blog.codinghorror.com/flattening-arrow-code/).